### PR TITLE
Compute an uncompressed digest for chunked layers

### PR DESCRIPTION
**Pull request by mtrmac**
… to ensure we can always use traditional layer IDs and image IDs, and use the uncompressed digests to avoid zstd:chunked view ambiguity, in c/image.

<s>Absolutely untested.</s>

### CODE REVIEW

File: pkg/chunked/storage_linux.go [hunk 1]
```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
...
+			}
+			if e.Digest == "" {
+				return nil, fmt.Errorf("missing digest for %q", e.Name)
+			}
+			digest, err := digest.Parse(e.Digest)
+			if err != nil {
+				return nil, err
+			}
+			d := digest.Encoded()
+			pathMap[filepath.Clean(e.Name)] = fmt.Sprintf("%s/%s", d[0:2], d[2:])
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:1775]: The file naming logic should not be duplicated.

File: pkg/chunked/storage_linux.go [hunk 2]
```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
...
+		}
+	}
+
+	dirFD2, err := unix.Dup(dirFD)
+	if err != nil {
+		return nil, err
+	}
+	rootDir := os.NewFile(uintptr(dirFD2), "staging root directory")
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:1783]: This is inelegant, maybe the `*os.File` creation can happen in the caller, making _that_ the primary holder, and the FD integer a derived value? At least the `Dup` should be avoidable.

File: pkg/chunked/storage_linux.go [hunk 3]
```diff
@@ -1609,6 +1615,20 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
...
 
+	if output.UncompressedDigest == "" && output.TarSplit != nil {
+		metadata := tsStorage.NewJSONUnpacker(bytes.NewReader(output.TarSplit))
+		fg, err := newStagedFileGetter(dirfd, differOpts, entriesWithOriginalNames)
+		if err != nil {
+			return output, fmt.Errorf("preparing stagedFileGetter: %w", err)
+		}
+		defer fg.Close()
+		digester := digest.Canonical.Digester()
+		if err := asm.WriteOutputTarStream(fg, metadata, digester.Hash()); err != nil {
```
Comment (cgwalters) [inline pkg/chunked/storage_linux.go:1626]: This is the expensive part where we're basically re-synthesizing the tarball from the unpacked files and the tar split-data just to get the checksum, right?

I know this is just a PoC but it seems very much worth a comment at the top of this conditional, something like:

```
// If we don't yet have the uncompressed digest, compute it now. This is needed
// for zstd:chunked for example to ensure callers identifying images by this
// have the data.
```

File: pkg/chunked/storage_linux.go [hunk 4]
```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
...
+			digest, err := digest.Parse(e.Digest)
+			if err != nil {
+				return nil, err
+			}
+			d := digest.Encoded()
+			pathMap[filepath.Clean(e.Name)] = fmt.Sprintf("%s/%s", d[0:2], d[2:])
+		}
+	}
+
+	dirFD2, err := unix.Dup(dirFD)
```
Comment (cgwalters) [inline pkg/chunked/storage_linux.go:1779]: This doesn't have `O_CLOEXEC` set

File: pkg/chunked/storage_linux.go [hunk 5]
```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
...
+
+func (fg *stagedFileGetter) Get(filename string) (io.ReadCloser, error) {
+	if fg.pathMap != nil {
+		path, ok := fg.pathMap[filepath.Clean(filename)]
+		if !ok {
+			return nil, fmt.Errorf("no path mapping exists for tar entry %q", filename)
+		}
+		filename = path
+	}
+	pathFD, err := securejoin.OpenatInRoot(fg.rootDir, filename)
```
Comment (cgwalters) [inline pkg/chunked/storage_linux.go:1806]: (A little surprising securejoin doesn't have a high level `securejoin.OpenReader` or something)

File: pkg/chunked/storage_linux.go [hunk 6]
```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
 
 	return digester.Digest() == digest
 }
+
+func newStagedFileGetter(dirFD int, differOpts *graphdriver.DifferOptions, entriesWithOriginalNames []fileMetadata) (*stagedFileGetter, error) {
```
Comment (cgwalters) [inline pkg/chunked/storage_linux.go:1758]: I don't know this codebase well but it seems surprising to me that we wouldn't have an equivalent of this function somewhere else, there has to already be code that's serializing zstd:chunked to tar in another place which has to do the same thing?

File: pkg/chunked/storage_linux.go (hunk 3 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go:1626]: > This is the expensive part where we're basically re-synthesizing the tarball from the unpacked files and the tar split-data just to get the checksum, right?

Yes.

File: pkg/chunked/storage_linux.go (hunk 6 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go:1758]: There is a getter in `layers.go` which doesn’t support the composefs digest mapping (in that case we fall back to a different expensive way to generate the diff, IIRC).

Comment (mtrmac) [reply pkg/chunked/storage_linux.go:1758]: I do agree that having those `FileGetter` implementations scattered around the codebase suggest that a better design might be possible. OTOH the path/digest mapping is expensive to maintain in a ready-to-use form.

Comment (cgwalters) [reply pkg/chunked/storage_linux.go:1758]: > OTOH the path/digest mapping is expensive to maintain in a ready-to-use form.

On the composefs side that's basically what the composefs blob is (plus all inline metadata) and it's pretty cheap to read back a deserialized abstract tree from a composefs blob, and a big bonus is that because fsverity is enabled on that file, validation is automatic.

Comment (mtrmac) [reply pkg/chunked/storage_linux.go:1758]: … yes, that’s a great point. There already is a special `overlayFileGetter` with composefs support, but it uses a temporary mount instead of reading the data directly. 

It seems likely that reading the data in user-space would be better, and that could be reused here.

But that’s, also, not the focus of this PR.

Comment (mtrmac) [reply pkg/chunked/storage_linux.go:1758]: Cc: @giuseppe

Comment (cgwalters) [reply pkg/chunked/storage_linux.go:1758]: It's also possible to deserialize a composefs in userspace via `composefs-info dump`; for the use case of turning "composefs into tar" that can be done without mounting it at a kernel level if desired. We have Rust code wrapping that in our crates. I suspect what would be helpful is to try to create a more extensive Go library for composefs separate from this project that fleshes things out like this.

Basically a cool thing about composefs is that it can be natively kernel mounted and provide full verity and things like that, but also because it's just metadata you can think of it much like a tar-split that can be handled in userspace too and efficiently merged etc.

File: pkg/chunked/storage_linux.go [hunk 7]
```diff
@@ -1687,6 +1705,30 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
...
+	// to one UncompressedDigest value for each layer, and that will avoid the ambiguity (in consumers who validate layers against DiffID).
+	//
+	// c/image also uses the UncompressedDigest as a layer ID, allowing it to use the traditional layer and image IDs.
+	//
+	// This is, sadly, quite costly: Up to now we might have only have had to write, and digest, only the new/modified files.
+	// Here we need to read, and digest, the whole layer, even if almost all of it was already present locally previously.
+	//
+	// Also, layers without a tar-split (estargz layers and old zstd:chunked layers) can't produce an UncompressedDigest that
+	// matches the expected RootFS.DiffID; they will need to use a special TOC-based layer ID, and we need to somehow (??)
+	// ensure unambiguity - either by always falling back to full pulls, or by never falling back.
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:1721]: @giuseppe @cgwalters note this.

zstd:chunked layers have tar-split in the current format since https://github.com/containers/storage/commit/b5413c2bd64742a08384300f09616a66231b21d1 , i.e. c/storage 1.54 == Podman 5.1.10.

I’m not 100% sure what to do here. I’m leaning towards outright refusing to partially pull estargz and old chunked layers (falling back to full pull), that is “clearly correct” and always possible. (Users who insist on always using partial pulls, e.g. for composefs, can rebuild their images.) Alternatively, we could always do a partial pull and refuse a fallback, but then we would need to break pulling those layers on other graph drivers (or, eventually, implement partial pulls for those graph drivers) — and we still would have the signing ambiguity.

**Review by mtrmac — COMMENTED** [review]
Updated:
- Introduced `composefs.RegularFilePathForValidatedDigest(d)` to centralize the path mapping logic
- Avoided the `Dup()` of the top-level directory file handle
- Added more comments.

@giuseppe RFC. Still mostly untested, but closer to the final form, I think.

File: pkg/chunked/storage_linux.go (hunk 1 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go]: Introduced `pkg/chunked/internal/composefs.RegularFilePathForValidatedDigest` for that.

File: pkg/chunked/storage_linux.go (hunk 2 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go]: Introduced `dirFile`.

(Eventually, it seems to me that it might be clean to have a `stagingDestination struct { dirFile , …}` holder, with various functions like `openDestinationFile` becoming methods on this new type, instead of passing a raw file descriptor as a parameter all over the place. That would clearly be a separate PR.)

File: pkg/chunked/storage_linux.go (hunk 3 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go:1726]: Comment added, discussing more of the goals, costs and tradeoffs.

File: pkg/chunked/storage_linux.go (hunk 4 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go]: This `Dup` is no longer necessary.

File: pkg/chunked/storage_linux.go (hunk 5 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go:1880]: (We have an `userns.secureOpen` internally with two other callers — centralizing that, and proposing that upstream, does sound attractive.)

**Comment by giuseppe** [comment]
changes LGTM

Anything more you'd like to do before it is ready for review/merge?

**Comment by mtrmac** [comment]
This could be merged as is, but the plan of record is to add an opt-out option.

But, most importantly, I want to test that this, along with containers/image#2613 , actually works.

File: pkg/chunked/storage_linux.go (hunk 7 above)
Comment (mtrmac) [reply pkg/chunked/storage_linux.go]: estargz and old zstd:chunked layers now fall back to full pulls by default, unless the user opts out of layer integrity enforcement.

**Comment by mtrmac** [comment]
Updated:
- estargz and old zstd:chunked layers (with no tar-split) fall back to full pulls
- Added an `"insecure_allow_unpredictable_image_contents"` `storage.conf` option to opt out of the layer consistency enforcement. Suggestions on better names (espcially more scary names) welcome. This also allows estargz  and old zstd:chunked layers again. The documentation intentionally does not advertise the performance benefit.
- Moved `RegularFilePathForValidatedDigest` into the just-added `pkg/chunked/internal/path` package.

Still untested ATM.

@giuseppe PTAL for an early review.

**Comment by mtrmac** [comment]
Temporarily includes #2197.

**Comment by mtrmac** [comment]
This PR is now ready for review / merging.

Testing in https://github.com/containers/podman/pull/24419#issuecomment-2573977213 (not yet automated) confirms that this PR works as expected.

@giuseppe PTAL

File: docs/containers-storage.conf.5.md [hunk 8]
```diff
@@ -124,6 +124,27 @@ The `storage.options.pull_options` table supports the following keys:
   It is an expensive operation so it is not enabled by default.
   This is a "string bool": "false"|"true" (cannot be native TOML boolean)
 
+**insecure_allow_unpredictable_image_contents="false"|"true"**
+  This should _almost never_ be set.
+  It allows partial pulls of images without guaranteeing that "partial
+  pulls" and non-partial pulls both result in consistent image contents.
```
Comment (mtrmac) [inline docs/containers-storage.conf.5.md:130]: BTW note the phrasing: without _guaranteeing_ consistency. I.e. this disables the _computation_ of uncompressed digest on partial pulls.

But images are still required to set correct DiffIDs: If we know the uncompressed digest for some reason (e.g. from a non-partial pull of the same blob, when the annotations were missing or the registry did not support partial pulls), and we see that it does not match DiffIDs, we are going to fail the pull.

**Comment by giuseppe** [comment]
@rhatdan @cgwalters PTAL
